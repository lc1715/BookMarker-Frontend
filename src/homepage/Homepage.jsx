import { useState, useEffect } from 'react';
import LoadSpinner from '../common/LoadSpinner';
import BookList from '../books/BookList';
import getNYTBooks from '../common/loadNYTBooks'
import useLocalStorage from '../hooks/useLocalStorage';


/**Homepage component
 * 
 * - calls getNYTBooks function to get nytbooks and checks if nytbooks contain new nyt books and timestamp
 *   or the stored nyt books from local storage
 * - if new nyt books and time stamp are returned, store them into local storage
 * - passes nytbooks and label for nyt books as props to BookList component
 * 
 * Route: '/'
 */
function Homepage() {
    const [nytBooks, setNytBooks] = useState(null);
    const [nytBookList, setNytBookList] = useLocalStorage('nytBooks');
    const [nytTimestamp, setNytTimestamp] = useLocalStorage('nytBooksTimestamp');

    // Runs at first render to get nyt book list and timestamp
    useEffect(() => {
        async function loadNYTBooks() {
            // gets stored nytbooks or new nytbooks and timestamp
            let bookList = await getNYTBooks();

            // check if new nyt book list and current timestamp exists and add to local storage
            if (bookList?.newBooks && bookList.currentTimeStamp) {
                let newBooks = bookList.newBooks;
                let currentTimeStamp = bookList.currentTimeStamp;

                // check if nyt books is empty
                if (newBooks.length === 0) {
                    setNytBooks(newBooks);
                    return;
                }
                // add new nyt books and current time stamp into local storage
                setNytBookList(JSON.stringify(newBooks));
                setNytTimestamp(currentTimeStamp);

                // set new nyt books to state
                setNytBooks(newBooks);
            } else {
                // set stored nyt books to state
                setNytBooks(bookList?.storedBooks);
            }
        };

        loadNYTBooks();
    }, []);

    if (!nytBooks) return <LoadSpinner />;

    return (
        <div>
            <BookList books={nytBooks} booksLabel={'NYT Bestsellers'} />
        </div>
    )
};

export default Homepage;
