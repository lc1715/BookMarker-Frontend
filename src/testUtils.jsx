import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
