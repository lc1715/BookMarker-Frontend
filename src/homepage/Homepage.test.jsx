import { it } from 'vitest';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../testUtils';
import Homepage from './Homepage';

it('renders the Homepage component', () => {
    render(
        <MemoryRouter>
            <Homepage />
        </MemoryRouter>
    );
})

it('matches snapshot when user is logged in', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <Homepage />
            </UserProvider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})

it('matches snapshot when user is logged out', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Homepage />
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})