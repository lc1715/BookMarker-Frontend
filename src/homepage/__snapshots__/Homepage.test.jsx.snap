// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot when user is logged in 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1lx8dk"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-lr6t8-MuiTypography-root"
    >
      <i
        class="fa-solid fa-spinner fa-spin-pulse fa-lg"
      />
      <br />
      Loading
    </p>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when user is logged out 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1lx8dk"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-lr6t8-MuiTypography-root"
    >
      <i
        class="fa-solid fa-spinner fa-spin-pulse fa-lg"
      />
      <br />
      Loading
    </p>
  </div>
</DocumentFragment>
`;
